share/lib/python/neuron/hclass3.py

```diff
@@ -4,38 +4,116 @@
 # h.anyclass methods may be overridden. If so the base method can be called
 # using the idiom self.basemethod = self.baseattr('methodname')
 # ------------------------------------------------------------------------------
-      
-from neuron import h, hoc
+
+__all__ = ["hclass", "nonlocal_hclass", "HocBaseObject"]
+
+from . import h, hoc
 import nrn
+import sys
+
+def assert_not_hoc_composite(cls):
+    """
+    Asserts that a class is not directly composed of multiple HOC types.
+    """
+    hoc_bases = [b for b in cls.__bases__ if issubclass(b, hoc.HocObject)]
+    if len(hoc_bases) > 1:
+        bases = ", ".join(b.__name__)
+        cname = cls.__name__
+        raise TypeError(f"Composition of {bases} HocObjects not allowed in {cname}")
+
+def hclass(hoc_type):
+    """
+    Class factory for subclassing HOC types.
+
+    Example:
+
+    ..code-block:: python
 
-#avoid syntax error if compiled by python 2
-exec('''
-class MetaHocObject(type):
-  """Provides Exception for Inheritance of multiple HocObject"""
-  def __new__(cls, name, bases, attrs):
-    #print cls, name, bases
-    m = []
-    for b in bases:
-      if issubclass(b, hoc.HocObject):
-        m.append(b.__name__)
-    if (len(m) > 1):
-      raise TypeError('Multiple Inheritance of HocObject in %s' % name
-        + ' through %s not allowed' % ','.join(m))
-      #note that join(b.__name__ for b in m) is not valid for Python 2.3
-  
-    return type.__new__(cls, name, bases, attrs)
-
-def hclass(c):
-    """Class factory for subclassing h.anyclass. E.g. class MyList(hclass(h.List)):..."""
-    if c == h.Section :
+        import neuron
+
+        class MyVector(neuron.hclass(neuron.h.Vector)):
+            pass
+
+    :param hoc_type: HOC types/classes such as ``h.List``, ``h.NetCon``, ``h.Vector``, ...
+    :type hoc_type: :class:`hoc.HocObject`
+    :deprecated: Inherit from :class:`~neuron.HocBaseObject` instead.
+    """
+    if hoc_type == h.Section:
```

Comment:
what happens if you call hclass on a subclass of a NEURON object?

Reply:
Could you give me some examples of what you mean with subclasses of NEURON objects?

Reply:
You may have already handled this, but I'm specifically thinking of:

    class MySection(nrn.Section):
        pass

    class YourSection(hclass(MySection)):
        pass

Reply:
Well since `nrn.Section` is not a subclass of `hoc.HocObject` and because there was this explicit bailout present in the code already to return `nrn.Section` from `hclass` if the input was `h.Section`, calling `hclass(nrn.Section)` or `hclass(MySection)` will raise a `TypeError` as there is no HOC type to create `HocObject`s from. 

Do you happen to know why this exception was introduced? Is `hoc.HocObject.__new__(*args, hocbase=h.Section)` inherently problematic?

Reply:
Let's talk more about this on Slack